### ISSUE TITLE
variance reduction techniques

### ISSUE THREAD

**Issue by dustinvtran** [blei-lab/edward#4]
- [x] use analytic form of q's entropy when available
- [x] use analytic form of KL(q(z) || p(z)) when available
- [ ] rao blackwellization
- [x] control variates (the vanilla one, nothing fancy yet)

**Comment by RJT1990** [comment]
Antithetic variables could be another option if the approx distribution is symmetric?

**Comment by dustinvtran** [comment]
Also a possibility! Antithetic variables are common in Monte Carlo, less so in variational inference. This means how it should be built, like control variates, should be general-purpose and not tied to a particular inference class.

**Comment by bhargavvader** [comment]
This still relevant? Any tips/papers on how to get at this?

**Comment by dustinvtran** [comment]
Apologies for the delay! Yes, it's very relevant. :) 

Here's my current thoughts on how to implement these:

1. __analytic KL.__ use the KL operator that some Google collaborators have been working on [[source]](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/distributions/python/ops/kullback_leibler.py) We can check if this operator is defined over pairs of random variables and apply it if it is.

2.  __rao blackwellization.__ this requires adding more variational inference infrastructure, so that we can update individual parameters in variational inference. (similar to how we manually update the parameters in the monte carlo classes)

3. __control variates.__ all easy to add. my suggestion would be to define utility functions outside the inference classes, and then call the function(s) during inference.

**Comment by bhargavvader** [comment]
Thanks for the response! I'll look into this.

**Comment by dustinvtran** [comment]
I have a WIP in the [`feature/rao-blackwellization`](https://github.com/blei-lab/edward/commits/feature/rao-blackwellization) branch. For a random variable `x_i`, it Rao-Blackwellizes the gradient with respect to other random variables in `latent_vars` and `data`. 

However, it doesn't Rao-Blackwellize with respect to the random variable's own `batch_shape`., i.e., `x_{-i}`. I'm not sure if there's enough information in `tf.contrib.distributions` to do this though. (For example, say we define `x = Normal(mu=tf.zeros(2), sigma=tf.ones(2))`. Can we tell that first element of `x` is parameterized by the first zero and one?

**Comment by lbollar** [comment]
I was interested in taking on the control variates piece. Best to create the utility function in separate module inside inference directory then?

Thanks!

**Comment by dustinvtran** [comment]
That sounds right. Then we can apply the utility function as a control variate for many methods.

**Comment by lbollar** [comment]
Apologies, I am trying to work this out but I think I have a lack of knowledge on how to do this inside the tensorflow computation graph. The calculations and methods are all very easy to understand, but I am struggling on how to handle everything inside the graph. 

Using the notation in [BBVI Paper](http://www.cs.columbia.edu/~blei/papers/RanganathGerrishBlei2014.pdf), i would need to calculate the covariance(f_d, h_d) and variance(h_d). Are there helper or utility functions that exist for this? 

Obviously I am lacking some knowledge and will continue to try, but I don't want to block this for anyone that can do it quicker. So if anyone wants to work on this simultaneously, feel free. Sorry, still getting used to working with tensorflow graph framework.

Thanks again

**Comment by dustinvtran** [comment]
@lbollar | The covariance and variance are calculated over (function evaluations of the) samples. This should be easy to calculate from the definitions for (co)variance. You can also try using something like [`tf.nn.moments`](https://www.tensorflow.org/versions/r0.10/api_docs/python/nn/normalization#moments) for the variance. You can ignore the "_d" subscripts and work purely on the full space (f, h) rather than the Rao-Blackwellized gradients. Hope that helps.

**Comment by lbollar** [comment]
@dustinvtran That helps a lot, thank you. I think I have something about ready for a working review. One question, I am having trouble finding it, but I thought I read something regarding control variates in that if you use the same sample to calculate `a = cov(f,h) / var(h)` that this result in a more biased estimate. Is this correct? Is it worth the trouble of having a second helper function that would use a different sample to calculate `a` if so, or is this just overkill?
